File: haskell/collatz-conjecture/src/CollatzConjecture.hs
module CollatzConjecture (collatz) where

collatz :: Integer -> Maybe Integer
collatz n = if n <= 0 then Nothing else step 0 n

step :: Integer -> Integer -> Maybe Integer
step s n | n == 1    = Just s
         | even n    = step (succ s) (div n 2)
         | otherwise = step (succ s) (n*3 + 1)

File: c/collatz-conjecture/src/collatz_conjecture.c
#include "collatz_conjecture.h"

static int collatz(int step, int n)
{
  return n == 1 ? step : collatz(step + 1, (n & 1) ? n * 3 + 1 : n / 2);
}

int steps(int start)
{
  return start <= 0 ? ERROR_VALUE : collatz(0, start);
}
